### ISSUE TITLE
trial warning reporting uses print to stdout

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2820]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2820|
|Type|defect|
|Created|2007-09-06 12:13:23Z|
|Branch|https://github.com/twisted/twisted/tree/leaky-warnings-2820|

See twisted/trial/runner.py:780-784

Using print means that the warning might not actually be reported, since print can fail with EINTR.  It also means that what was once structured information (this is a warning, this is the message format string, these are the arguments) becomes unstructured data in a log file; sometimes this is desirable, but it should probably be up to a reporter what is done with the message, not the runner.

<details><summary>Searchable metadata</summary>

```
trac-id__2820 2820
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_leaky_warnings_2820 branches/leaky-warnings-2820
branch_author__jonathanj jonathanj
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1189080803000000 1189080803000000
changetime__1249911855000000 1249911855000000
version__None None
owner__jonathanj jonathanj

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

See duplicate [#3070](https://github.com/twisted/twisted/issues/3070)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

See related ticket [#2365](https://github.com/twisted/twisted/issues/2365)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[#3433](https://github.com/twisted/twisted/issues/3433) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [25279]) Merge flushwarnings-3487

Author: radix, itamar, exarkun
Reviewer: glyph
Fixes: [#3487](https://github.com/twisted/twisted/issues/3487)
Fixes: [#3427](https://github.com/twisted/twisted/issues/3427)
Fixes: [#2820](https://github.com/twisted/twisted/issues/2820)
Fixes: [#3506](https://github.com/twisted/twisted/issues/3506)

Introduce `TestCase.flushWarnings`, an API for interacting with the Python warnings
module in unit tests.  By default, the default trial reporter will write warnings
emitted by tests to its output stream (no longer does the runner print them to
stdout).  `flushWarnings` may be used to prevent this from happening in particular
tests, and its return value may be used to make assertions about what warnings have
been emitted.  `TestCase.assertWarns` and `TestCase.callDeprecated` are both now
implemented in terms of this more flexible API.

The default reporter includes rudamentary support for duplicate suppression: each
warning will be reported only once per test, but if a warning is emitted by two or
more tests, it will be reported for each test (isolating tests from each other
further by making warning behavior the same regardless of what other tests have run).

There is also some support for failing tests based on warnings emitted.  If the Python
warnings filters system is used to turn a warning into an exception, any test which
causes such a warning to be emitted will have an error attributed to it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `reopened`|
|-|-|

Apparently this wasn't really fixed. :(

```
exarkun@charm:~$ trial twisted.test.test_log.LogTest.test_showwarning
twisted.test.test_log
  LogTest
    test_showwarning ... warning-filename.py:27: FakeWarning: unique warning message
                                                  [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 0.002s

PASSED (successes=1)
exarkun@charm:~$ 
```

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj commented|
|-|-|

(In [27219]) Branching to 'leaky-warnings-2820'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj commented|
|-|-|

(In [27220]) Make log.showwarning a method of LogPublisher and use a private log publisher
for test_showwarning, to avoid other observers from polluting stdout,
refs [#2820](https://github.com/twisted/twisted/issues/2820).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj removed owner<br>@jonathanj set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jonathanj|
|-|-|

Okay, great.  It looks like this ticket was mostly resolved by the previous branch.  Only the tests needed the slight tweak in this new branch to avoid spitting ugly crap into the output.  The change makes sense and looks good.  Please update the copyright date on both modified files and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_12"></a>|@jonathanj commented|
|-|-|

(In [27224]) Update copyright dates, refs [#2820](https://github.com/twisted/twisted/issues/2820).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_13"></a>|@jonathanj commented|
|-|-|

(In [27225]) Update the other copyright date, refs [#2820](https://github.com/twisted/twisted/issues/2820).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_14"></a>|@jonathanj set status to `closed`|
|-|-|

(In [27226]) Merge leaky-warnings-2820 branch.

Author: jonathanj
Revewer: exarkun
Fixes: [#2820](https://github.com/twisted/twisted/issues/2820)

Tweak the tests for "showwarning" to avoid emitting a log message to all the
normal observers, some of which simply print warnings to stdout, causing ugly
spew in trial's output.
